_src/_posts/2017-01-31-learning-rtb.markdown

```diff
@@ -0,0 +1,204 @@
+events of interest (clicks, installs, opens, etc.). A conversion rate for an
+event $E$ is the conditional probability $p(E|C,T)$ of getting that event by
+printing some creative $C$ in the context of the current RTB transaction $T$.
+Why is it so important for our business to have good estimates of these rates?
+Simply because we sell conversions but we pay for impressions. For example,
+suppose one client pays us $\$1$ per install and we estimate the install
+conversion rate to be around $0.1$ (just cheating... real world figures are well
+under a dismaying $0.001$ `:(`). Then, assuming our creative is going to get
+printed, we should expect an average profit of $\$0.1 - c$, where $c$ is the
+expected cost of the impression. If our rate estimates are grossly miscalculated
+we will be losing money by valuating the expected income too low (thus missing
+opportunities with $c < \$0.1$) or too high (thus buying non-opportunities with
+$c > \$0.1$)[^risk]. The bottom line is that our bidder may be able to pick an
+optimal pair (bid, creative) for the market transaction in course from:
+(i) The estimated conditional conversion rate for the goal event.
+(ii) Our customer valuation of the goal event.
+(iii) The budget constraint for the advertising campaign.
+(iv) A narrow-minded profit maximization behavior.
+
+In order to estimate conversion rates, we implemented a SGD (Stochastic
+Gradient Descent) algorithm or, more specifically, [a FTRL-P][ftrlp] (Follow the
+Regularized Leader Proximal) algorithm with [Adagrad][adagrad] (Adaptive
+Learning Rate) and [ElasticNet][enet] (Lasso and Ridge) regularization. This is
+a state-of-the-art algorithm mostly used in online advertising because of its
+[convenient trade-off][sgd] between optimization and estimation errors in
+scenarios where computational capacity ---and not available data--- is the
+real bottleneck. Every day our system is able to learn from an online stream of
+tens of millions of events published by the bidder into a very
+lightweight middleware implemented on top of [ZMQ][zmq].
+
+The first step after receiving an event from the stream is to turn it into a
+numerical vector that the estimator can consume. Since most of our variables are
+categorical and since we consider many interactions between these variables, the
+output vector will be a sparse binary array of "dummy" variables. We feed this
+vector to a cluster of estimators, each one parameterized according to a grid of
+parameters controlling strength of regularization, number of hashing bits (yes,
+we take advantage of the [hashing trick][trick] as everybody else does),
+estimator memory, etc. Periodically, we evaluate the estimators in the cluster
+using a mix of progressive validation and bootstrap validation in order to
+select the best performing parameterization. The chosen estimator is then
+converted to a tree-like data structure thoroughly optimized for fast model
```

Comment:
If you want you could go into a little more detail about the serialization/deserialization method.